src/util/secure_socket_thread.c

```diff
@@ -0,0 +1,211 @@
+
+    charbuf data;
+    charbuf output;
+
+    //Parse request for processing
+    if (request_decoder(request, &request_type, &key_id, &data_in, &request_sig, &requestor_cert))
+    {
+      err_message = "Missing Data";
+      error_message_encoder(&message, err_message);
+      pelz_log(LOG_DEBUG, "%d::Error: %.*s, %d", new_socket, (int) message.len, message.chars, (int) message.len);
+      pelz_key_socket_close(new_socket);
+      free_charbuf(&request);
+      return;
+    }
+
+    free_charbuf(&request);
+
+    decodeBase64Data(data_in.chars, data_in.len, &data.chars, &data.len);
+    free_charbuf(&data_in);
+
+    pthread_mutex_lock(&lock);
+    pelz_request_handler(eid, &status, request_type, key_id, data, &output);
+    if (status == KEK_NOT_LOADED)
+    {
+      if (key_load(key_id) == 0)
+      {
+        pelz_request_handler(eid, &status, request_type, key_id, data, &output);
+      }
+      else
+      {
+        status = KEK_LOAD_ERROR;
+      }
+    }
+    pthread_mutex_unlock(&lock);
+    free_charbuf(&data);
+
+    if (status != REQUEST_OK)
+    {
+      pelz_log(LOG_ERR, "%d::Service Error\nSend error message.", new_socket);
+      switch (status)
+      {
+      case KEK_LOAD_ERROR:
+        err_message = "Key not added";
+        break;
+      case KEY_OR_DATA_ERROR:
+        err_message = "Key or Data Error";
+        break;
+      case ENCRYPT_ERROR:
+        err_message = "Encrypt Error";
+        break;
+      case DECRYPT_ERROR:
+        err_message = "Decrypt Error";
+        break;
+      case REQUEST_TYPE_ERROR:
+        err_message = "Request Type Error";
+        break;
+      case CHARBUF_ERROR:
+        err_message = "Charbuf Error";
+        break;
+      default:
+        err_message = "Unrecognized response";
+      }
+      error_message_encoder(&message, err_message);
+    }
+    else
+    {
+      encodeBase64Data(output.chars, output.len, &data_out.chars, &data_out.len);
+      if (strlen((char *) data_out.chars) != data_out.len)
+      {
+        data_out.chars[data_out.len] = 0;
+      }
+      message_encoder(request_type, key_id, data_out, &message);
+      pelz_log(LOG_DEBUG, "%d::Message Encode Complete", new_socket);
+      pelz_log(LOG_DEBUG, "%d::Message: %.*s, %d", new_socket, (int) message.len, message.chars, (int) message.len);
+    }
+    free_charbuf(&key_id);
+    free_charbuf(&data_out);
+    free_charbuf(&output);
+
+    pelz_log(LOG_DEBUG, "%d::Message & Length: %.*s, %d", new_socket, (int) message.len, message.chars, (int) message.len);
+    //Send processed request back to client
+    if (pelz_key_socket_send(new_socket, message))
+    {
+      pelz_log(LOG_ERR, "%d::Socket Send Error", new_socket);
+      free_charbuf(&message);
+      while (!pelz_key_socket_check(new_socket))
+      {
+        continue;
+      }
+      pelz_key_socket_close(new_socket);
+      return;
+    }
+    free_charbuf(&message);
+  }
+  pelz_key_socket_close(new_socket);
+  return;
+}
+
```

Comment:
I think it would be better to start out by reusing the unsecure_socket_thread code as much as possible, and gradually break it up and replace it as needed.
Copying the code all at once will probably lead to unnecessary duplication.